ethcore/light/src/provider.rs

```diff
@@ -226,11 +226,76 @@ impl<T: ProvingBlockChainClient + ?Sized> Provider for T {
 		self.code_by_hash(req.account_key, BlockId::Hash(req.block_hash))
 	}
 
-	fn header_proof(&self, _req: request::HeaderProof) -> Option<(encoded::Header, Vec<Bytes>)> {
-		None
+	fn header_proof(&self, req: request::HeaderProof) -> Option<(encoded::Header, Vec<Bytes>)> {
+		use util::MemoryDB;
+		use util::trie::{Trie, TrieMut, TrieDB, TrieDBMut, Recorder};
+
+		if Some(req.cht_number) != ::cht::block_to_cht_number(req.block_number) {
+			debug!(target: "les_provider", "Requested CHT number mismatch with block number.");
+			return None;
+		}
+
+		let mut memdb = MemoryDB::new();
+		let mut root = H256::default();
+		let mut needed_hdr = None;
+		{
+			let mut t = TrieDBMut::new(&mut memdb, &mut root);
+			let start_num = ::cht::start_number(req.cht_number);
+			for i in (0..::cht::SIZE).map(|x| x + start_num) {
+				match self.block_header(BlockId::Number(i)) {
+					None => return None,
+					Some(hdr) => {
+						t.insert(
+							&*::rlp::encode(&i),
+							&*::rlp::encode(&hdr.hash()),
+						).expect("fresh in-memory database is infallible; qed");
+
+						if i == req.block_number { needed_hdr = Some(hdr) }
+					}
+				}
+			}
+		}
+		let needed_hdr = needed_hdr.expect("`needed_hdr` always set in loop, number checked before; qed");
```

Comment:
it's guaranteed by the check that `req.cht_number` is equal to the calculated CHT number based on the block number in the request. this means that the function has already returned in the case that the requested number wouldn't fall within the range of blocks we iterate.